http-client/Network/HTTP/Client/Connection.hs

```diff
@@ -85,15 +86,35 @@ makeConnection :: IO ByteString -- ^ read
                -> IO Connection
 makeConnection r w c = do
     istack <- newIORef []
+
+    -- it is necessary to make sure we never read from or write to
+    -- already closed connection.
+    closedVar <- newMVar False
```

Comment:
I'm concerned about adding `MVar` overhead for every operation. The way that the Manager system works, a `Connection` is only leased to a single thread at any given time, so therefore using an `IORef` should be cheaper and still thread-safe.